metrics_test.go

```diff
@@ -0,0 +1,326 @@
+package metrics
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/transcovo/go-chpr-logger"
+	"github.com/transcovo/go-chpr-metrics/utils/tests"
+)
+
+var port int = 12345
+var localhost = "127.0.0.1"
+
+/*
+Ensure unicity in port number (tests are ran in parallel)
+*/
+func getUniquePort() string {
+	port++
+	return strconv.Itoa(port)
+}
+
+/*
+Start an UDP server that does nothing. Waits for the test to finish to close the server
+Notify the channel when cleaning is done
+*/
+func simpleUDPServer(host, port string, end chan string) {
+	pc, err := net.ListenPacket("udp", host+":"+port)
+	if err != nil {
+		fmt.Println("Error creating the udp server", err)
+		panic(err)
+	}
+	end <- ""  // Notify that ready
+	<-end      // Wait for the test to notify that it's finishing
+	pc.Close() // We do not use defer here because we need to close cleanly the connection
+	end <- ""
+}
+
+/*
+Notify and wait for a answer
+*/
+func notifyToClose(end chan string) {
+	end <- ""
+	<-end
+}
+
+/*
+Test the defaultClientErrorHandler
+*/
+func TestDefaultClientErrorHandler(t *testing.T) {
+	err := errors.New("This is an error")
+
+	stdout := tests.CaptureStdout(func() { defaultClientErrorHandler(err) })
+	assert.Contains(t, stdout, `msg="[Metrics] Error caught in metrics" error="This is an error"`)
+}
+
+/**
+This tests expect buildClient to fail if the fields (like the port) in the config object are ill formatted
+*/
+func TestBuildClient_Failure(t *testing.T) {
+	wrongConfig := &clientConfig{
+		Host:   "an ill formatted host",
+		Prefix: "prefix.",
+	}
+
+	assert.Panics(t, func() {
+		defaultClientBuilder.buildClient(wrongConfig)
+	})
+}
+
+/**
+Tests that with a started udp server, the statsd client is correctly started
+*/
+func TestBuildClient_Success(t *testing.T) {
+	port := getUniquePort()
+	host := localhost + ":" + port
+
+	end := make(chan string)
+	go simpleUDPServer(localhost, port, end)
+	<-end
+	correctConfig := &clientConfig{
+		Host:   host,
+		Prefix: "prefix.",
+	}
+
+	mockClientBuilder := &clientBuilder{
+		errorLogger: func(err error) {
+			logger.WithField("error", err).Error("this is some error")
+		},
+	}
+
+	client := mockClientBuilder.buildClient(correctConfig)
+	assert.NotNil(t, client)
+	notifyToClose(end)
+}
+
+/*
+Tests that the configuration is built correctly from a metricsDestination object
+*/
+func TestClientConfigFromDestination_Success(t *testing.T) {
+	destination := &metricsDestination{
+		Host:   "some.host",
+		Port:   "1234",
+		Prefix: "prefix.",
+	}
+
+	config := clientConfigFromDestination(destination)
+	assert.Equal(t, config.Host, "some.host:1234")
```

Comment:
Ah cool ! I checked the code of assert.equal and it does a reflect.DeeEqual ! Did not know, cool stuff to learn !